### PR TITLE
[SDC] Integrate deterministic replay into Trainer

### PR DESCRIPTION
**Pull request by xmfan**
Stacked PRs:
 * __->__#4334

--- --- ---

### [SDC] Integrate deterministic replay into Trainer

Add opt-in deterministic replay for detecting silent data corruption during
training.

The Trainer builds an SDCReplayer when config.sdc_replay is set and holds it
as self.sdc_replayer (config.sdc_replay is None by default, which disables
replay). For each optimizer step, the step's first forward/backward call --
one gradient-accumulation group, which under pipeline parallelism is one
complete pipeline schedule -- is routed through sdc_replayer.run_fwd_bwd,
which replay-checks it when scheduled. Later accumulation groups exercise
the same compute and communication paths and run unchecked. A mismatch
raises SDCReplayMismatch on every rank before gradient clipping, optimizer,
scheduler, or checkpoint updates can observe corrupted state.

The replayer is built from config with narrow runtime inputs (modules,
device, and a ScalarStateAccessor for ntokens_seen, which the replayed
forward/backward mutates; self.step is incremented outside the replay
boundary and needs no capture); it does not receive the Trainer. train_step
enters the replay boundary right after optimizers.zero_grad, satisfying the
replayer's empty-entry-gradients contract. All replay
scheduling state lives inside the replayer; the check schedule is
process-local, not checkpointed, and restarts after checkpoint loading via
reset_schedule.

GraphTrainer shares the same replay boundary by inheriting Trainer.

The feature is disabled by default and has no CLI surface (sdc_replay is
suppressed from tyro parsing); enable it programmatically:

```python
from torchtitan.observability.sdc_replayer import SDCReplayer

config.debug.deterministic = True
config.sdc_replay = SDCReplayer.Config(
    num_steps=1,    # optimizer steps checked after each (re)start; -1 checks every step
    num_replays=1,  # re-executions compared against the reference
)
```

Replay requires debug.deterministic=True and
debug.deterministic_warn_only=False. Eager execution, torch.compile, CUDA
graphs, symmetric-memory FSDP, distributed GEMM, async TP, DeepEP v2,
HybridEP, and MinimalAsyncEP use the same replay boundary. CUDA graphs
require num_replays=1; this is validated purely from config (independent of
runtime CUDA graph availability) because additional replays would require
CUDA-graph-aware restoration of graph-owned gradient and optional-buffer
storage.

After config overrides are applied, Trainer re-runs the full
Config.__post_init__ (raise-only) instead of individual validators.

Document the feature under torchtitan/observability/ and add CPU and GPU
coverage including an intentionally corrupted distributed gradient. The
sdc_replay_mismatch integration test sets use_real_pg=True because
cross-rank mismatch propagation (all_reduce and all_gather_object) needs
real collectives.

- pytest -q tests/unit_tests/cpu/test_trainer.py
  tests/unit_tests/cpu/observability/test_sdc_replayer.py
  tests/unit_tests/cpu/test_invalid_loss.py
  tests/unit_tests/cpu/test_config_manager.py
- tests/integration_tests/features.py::sdc_replay_cudagraph
- torchtitan/experiments/graph_trainer/tests/integration_tests.py::sdc_replay
- tests/integration_tests/h100.py::deepseek_v3_fsdp+cp+tp+minimal_async_ep+sdc_replay
- tests/integration_tests/features.py::sdc_replay_mismatch
- git diff --check
- pre-commit run --all-files

### CODE REVIEW

File: docs/silent_data_corruption.md
Comment (tianyu-l) [inline docs/silent_data_corruption.md:1]: please put under `observability/`

File: tests/integration_tests/features.py [hunk 1]
```diff
@@ -17,6 +17,20 @@ def build_features_test_list() -> list[OverrideDefinitions]:
...
+            test_descr="SDC replay with CUDA graphs",
+            test_name="sdc_replay_cudagraph",
+            ngpu=1,
+            skip_rocm_test=True,
+        ),
+        OverrideDefinitions(
+            configs=[recipes.deepseek_v3_debugmodel_sdc_replay_mismatch],
+            test_descr="SDC replay detects a distributed gradient mismatch",
+            test_name="sdc_replay_mismatch",
+            ngpu=2,
```
Comment (tianyu-l) [inline tests/integration_tests/features.py:31]: @wwwjn should / how should we use this field after the CI refactor? I think the idea is that here we need one real pg test for sdc replay. Do you think this is needed?

File: tests/unit_tests/cpu/test_no_new_cli_options.py [hunk 2]
```diff
@@ -168,6 +168,9 @@
         "profiler.profiler_warmup",
         "profiler.save_memory_snapshot_folder",
         "profiler.save_traces_folder",
+        "sdc_replay.enabled",
+        "sdc_replay.num_replays",
+        "sdc_replay.num_steps",
```
Comment (tianyu-l) [inline tests/unit_tests/cpu/test_no_new_cli_options.py]: we're moving away from CLI args. For SDC let's suppress tyro parsing so that this file is not changed

File: torchtitan/trainer.py [hunk 3]
```diff
@@ -198,6 +203,30 @@ def _validate_cuda_graphs(self) -> None:
...
+                return
+            if not self.debug.deterministic:
+                raise ValueError("SDC replay requires debug.deterministic=True.")
+            if self.debug.deterministic_warn_only:
+                raise ValueError(
+                    "SDC replay requires debug.deterministic_warn_only=False."
+                )
+            if (
+                not self.training.disable_cuda_graphs
+                and is_cudagraph_available()
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: we should error out without this condition?

File: torchtitan/trainer.py [hunk 4]
```diff
@@ -198,6 +203,30 @@ def _validate_cuda_graphs(self) -> None:
...
+                and self.sdc_replay.num_replays > 1
+            ):
+                # TODO: Support additional replays after CUDA graph capture has
+                # established stable gradient and buffer identities, or make
+                # replay state restoration aware of graph-owned storage.
+                raise ValueError(
+                    "SDC replay with CUDA graphs requires "
+                    "sdc_replay.num_replays=1. Set --sdc-replay.num-replays 1 "
+                    "or --training.disable_cuda_graphs."
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: please make this message CLI free

File: torchtitan/trainer.py [hunk 5]
```diff
@@ -350,6 +380,7 @@ def __init__(self, config: Config):
         if config.override.imports:
             apply_overrides(config.override, config)
         config._validate_cuda_graphs()
+        config._validate_sdc_replay()
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: should we rerun the entire `__post_init__`

File: torchtitan/trainer.py [hunk 6]
```diff
@@ -748,6 +783,37 @@ def post_dataloading_process(
...
+        return execute()
+
+    def _get_sdc_replay_modules(self) -> Iterable[torch.nn.Module]:
+        return self.model_parts
+
+    def _get_sdc_replay(self) -> SDCReplay:
+        if self._sdc_replay is None:
+            self._sdc_replay = SDCReplay(
+                config=self.config.sdc_replay,
+                trainer=self,
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: shouldn't need to send the entire trainer to SDCReplay which sounds a sub component

File: torchtitan/trainer.py [hunk 7]
```diff
@@ -748,6 +783,37 @@ def post_dataloading_process(
         return inputs, labels, extra_kwargs
 
     @sl.log_trace_span("fwd_bwd")
+    def run_forward_backward(
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: looks like a lot sdc internal gets exposed in main trainer, can we do something like https://fburl.com/code/n7tvf7p6

File: torchtitan/trainer.py [hunk 8]
```diff
@@ -748,6 +783,37 @@ def post_dataloading_process(
...
+            sdc_replay = self._get_sdc_replay()
+            if sdc_replay.should_run(self.sdc_attempt_step):
+                return sdc_replay.run(
+                    execute,
+                    step=self.step,
+                    attempt=self.sdc_attempt_step + 1,
+                )
+        return execute()
+
+    def _get_sdc_replay_modules(self) -> Iterable[torch.nn.Module]:
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: I think we can remove this if we don't need to worry about other trainers like FLUX

File: torchtitan/models/flux/trainer.py [hunk 9]
```diff
@@ -283,10 +283,18 @@ def forward_backward_step(
 
         return loss
 
+    def _get_sdc_replay_modules(self) -> Iterable[torch.nn.Module]:
+        modules = [*self.model_parts, self.t5_encoder, self.clip_encoder]
+        if self.autoencoder is not None:
+            modules.append(self.autoencoder)
+        return modules
```
Comment (tianyu-l) [inline torchtitan/models/flux/trainer.py]: let's not worry about flux for now

File: torchtitan/experiments/torchft/trainer.py
Comment (tianyu-l) [inline torchtitan/experiments/torchft/trainer.py:1]: let's also not worry about ft

File: torchtitan/trainer.py [hunk 10]
```diff
@@ -748,6 +783,37 @@ def post_dataloading_process(
...
+        return self.model_parts
+
+    def _get_sdc_replay(self) -> SDCReplay:
+        if self._sdc_replay is None:
+            self._sdc_replay = SDCReplay(
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: should build from config

File: torchtitan/trainer.py [hunk 11]
```diff
@@ -271,6 +300,7 @@ def maybe_log(self) -> None:
     # additional training states
     step: int
     ntokens_seen: int
+    sdc_attempt_step: int
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: too vague .. do you mean `step_after_checkpoint_load`

File: torchtitan/trainer.py [hunk 12]
```diff
@@ -550,6 +581,10 @@ def __init__(self, config: Config):
         # These attributes must be initialized before checkpoint loading.
         self.step = 0
         self.ntokens_seen = 0
+        # Initialize process-local SDC replay state. This state is not checkpointed.
+        self.sdc_attempt_step = 0
+        self._sdc_replay_unit_index = 0
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: what is a "unit" and what's the "index" of a unit

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/trainer.py [hunk 13]
```diff
@@ -551,6 +584,21 @@ def __init__(self, config: Config):
...
 
+        # SDC replay state is process-local and not checkpointed; its check
+        # schedule restarts after every checkpoint load (see load_state_dict).
+        self.sdc_replayer = None
+        if config.sdc_replay.enabled:
+            self.sdc_replayer = config.sdc_replay.build(
+                modules=self.model_parts,
+                device=self.device,
+                scalar_state={
+                    "ntokens_seen": ScalarStateAccessor(
```
Comment (tianyu-l) [inline torchtitan/trainer.py:594]: how about `"step"`? I guess the idea is it's not error-prone?

File: torchtitan/observability/silent_data_corruption.md
Comment (tianyu-l) [inline torchtitan/observability/silent_data_corruption.md:1]: sorry to clarify: let's put this under torchtitan/observability/silent_data_corruption.md, as we plan to move all docs to local folders

File: docs/observability/silent_data_corruption.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+at least one, and higher values catch intermittent corruption a single replay
+can miss.
+
+Replay requires `debug.deterministic=True`,
+`debug.deterministic_warn_only=False`, and uses `torch.hash_tensor`.
+
+## What is replayed
+
+Only the first forward/backward of a checked optimizer step is replayed: one
+full gradient-accumulation group or, with pipeline parallelism, one complete
```
Comment (tianyu-l) [inline docs/observability/silent_data_corruption.md]: slightly inaccurate -- gradient accum can be used with or without PP, so even with PP it's replaying a full gradient accumulation step

File: docs/observability/silent_data_corruption.md [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+re-executed and compared against the initial reference execution; it must be
+at least one, and higher values catch intermittent corruption a single replay
+can miss.
+
+Replay requires `debug.deterministic=True`,
+`debug.deterministic_warn_only=False`, and uses `torch.hash_tensor`.
+
+## What is replayed
+
+Only the first forward/backward of a checked optimizer step is replayed: one
```
Comment (tianyu-l) [inline docs/observability/silent_data_corruption.md]: where's this limitation from -- why can't we replay multiple grad accum steps

**Review by tianyu-l — APPROVED** [review]
Looks very nice!

Only feel a bit mysterious on why we're limited to replay the first step of grad accumulation.

File: tests/integration_tests/features.py (hunk 1 above)
Comment (wwwjn) [reply tests/integration_tests/features.py:31]: `ngpu` set the world_size. For a single OverrideDefinitions test, it will be run as 1 GPU fake pg test when PR is created / synced, use `ngpu` to decided the worldsize it mimics.  And the same test will also be run on 8GPU real PG when the PR is merged.

If SDC is only runnable on real_pg, you should set `use_real_pg=True`

File: docs/observability/silent_data_corruption.md (hunk 15 above)
Comment (xmfan) [reply docs/observability/silent_data_corruption.md]: we can add it as an option, it's just more expensive to rerun

File: torchtitan/trainer.py (hunk 13 above)
Comment (xmfan) [reply torchtitan/trainer.py:594]: Sry for the confusion, so right now this exists because the fwd_bwd that is being replayed does `self.ntokens_seen += tokens` each iteration. SDC needs to reset the state in between capture and replay, but since `self.ntokens_seen` is not part of the model's nn.module state (as buffer registration), it's hard to reset it properly from outside. This is essentially a reset() mechanism, implemented via manual registration of the mutation.

`ntokens_seen` is for logging purposes only, so i understand why we don't want it in model code. but if we do add it to the module's buffers, SDC could handle it together with the module.buffers reset.

File: torchtitan/trainer.py [hunk 16]
```diff
@@ -112,6 +113,10 @@ class Config(Configurable.Config):
         comm: CommConfig = field(default_factory=CommConfig)
         validator: Validator.Config = field(default_factory=Validator.Config)
         debug: DebugConfig = field(default_factory=DebugConfig)
+        # NOTE: sdc_replay is suppressed from tyro CLI parsing; enable it
+        # programmatically in a config/recipe by assigning a config
+        # (config.sdc_replay = SDCReplayer.Config()). None disables replay.
+        sdc_replay: Annotated[SDCReplayer.Config | None, tyro.conf.Suppress] = None
```
Comment (tianyu-l) [inline torchtitan/trainer.py]:
```suggestion
        sdc_replayer: Annotated[SDCReplayer.Config | None, tyro.conf.Suppress] = None
```
